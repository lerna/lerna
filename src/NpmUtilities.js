import log from "npmlog";
import onExit from "signal-exit";
import path from "path";
import writePkg from "write-pkg";

import ChildProcessUtilities from "./ChildProcessUtilities";
import FileSystemUtilities from "./FileSystemUtilities";
import splitVersion from "./utils/splitVersion";

function execInstall(directory, {
  registry,
  npmClient,
  npmClientArgs,
  npmGlobalStyle,
  mutex,
}) {
  // build command, arguments, and options
  const opts = NpmUtilities.getExecOpts(directory, registry);
  const args = ["install"];
  let cmd = npmClient || "npm";

  if (npmGlobalStyle) {
    cmd = "npm";
    args.push("--global-style");
  }

  if (cmd === "yarn" && mutex) {
    args.push("--mutex", mutex);
  }

  if (cmd === "yarn") {
    args.push("--non-interactive");
  }

  if (npmClientArgs && npmClientArgs.length) {
    args.push(...npmClientArgs);
  }

  log.silly("installInDir", [cmd, args]);
  return ChildProcessUtilities.exec(cmd, args, opts);
}

export default class NpmUtilities {
  static installInDir(directory, dependencies, config, callback) {
    log.silly("installInDir", path.basename(directory), dependencies);

    // Nothing to do if we weren't given any deps.
    if (!(dependencies && dependencies.length)) {
      log.verbose("installInDir", "no dependencies to install");
      return callback();
    }

    const packageJson = path.join(directory, "package.json");
    const packageJsonBkp = packageJson + ".lerna_backup";

    log.silly("installInDir", "backup", packageJson);
    FileSystemUtilities.rename(packageJson, packageJsonBkp, (err) => {
      if (err) {
        log.error("installInDir", "problem backing up package.json", err);
        return callback(err);
      }

      const cleanup = () => {
        log.silly("installInDir", "cleanup", packageJson);
        // Need to do this one synchronously because we might be doing it on exit.
        FileSystemUtilities.renameSync(packageJsonBkp, packageJson);
      };

      // If we die we need to be sure to put things back the way we found them.
      const unregister = onExit(cleanup);

      // We have a few housekeeping tasks to take care of whether we succeed or fail.
      const done = (err) => {
        cleanup();
        unregister();
        callback(err);
      };

      // Construct a basic fake package.json with just the deps we need to install.
      const tempJson = {
        dependencies: dependencies.reduce((deps, dep) => {
          const [pkg, version] = splitVersion(dep);
          deps[pkg] = version || "*";
          return deps;
        }, {})
      };

      log.silly("installInDir", "writing tempJson", tempJson);
      // Write out our temporary cooked up package.json and then install.
      writePkg(packageJson, tempJson)
        .then(() => execInstall(directory, config))
        .then(() => done(), done);
    });
  }

  static installInDirOriginalPackageJson(directory, config, callback) {
    log.silly("installInDirOriginalPackageJson", directory);

    return execInstall(directory, config)
      .then(() => callback(), callback);
  }


  static addDistTag(directory, packageName, version, tag, registry) {
    log.silly("addDistTag", tag, version, packageName);

    const opts = NpmUtilities.getExecOpts(directory, registry);
    ChildProcessUtilities.execSync("npm", ["dist-tag", "add", `${packageName}@${version}`, tag], opts);
  }

  static removeDistTag(directory, packageName, tag, registry) {
    log.silly("removeDistTag", tag, packageName);

    const opts = NpmUtilities.getExecOpts(directory, registry);
    ChildProcessUtilities.execSync("npm", ["dist-tag", "rm", packageName, tag], opts);
  }

  static checkDistTag(directory, packageName, tag, registry) {
    log.silly("checkDistTag", tag, packageName);

    const opts = NpmUtilities.getExecOpts(directory, registry);
    return ChildProcessUtilities.execSync("npm", ["dist-tag", "ls", packageName], opts).indexOf(tag) >= 0;
  }

  static runScriptInDir(script, args, directory, callback) {
    log.silly("runScriptInDir", script, args, path.basename(directory));

    const opts = NpmUtilities.getExecOpts(directory);
    ChildProcessUtilities.exec("npm", ["run", script, ...args], opts, callback);
  }

  static runScriptInDirSync(script, args, directory, callback) {
    log.silly("runScriptInDirSync", script, args, path.basename(directory));

    const opts = NpmUtilities.getExecOpts(directory);
    ChildProcessUtilities.execSync("npm", ["run", script, ...args], opts, callback);
  }

  static runScriptInPackageStreaming(script, args, pkg, callback) {
    log.silly("runScriptInPackageStreaming", [script, args, pkg.name]);

    const opts = NpmUtilities.getExecOpts(pkg.location);
    ChildProcessUtilities.spawnStreaming(
      "npm", ["run", script, ...args], opts, pkg.name, callback
    );
  }

  static publishTaggedInDir(tag, directory, registry, callback) {
    log.silly("publishTaggedInDir", tag, path.basename(directory));

    const opts = NpmUtilities.getExecOpts(directory, registry);
    ChildProcessUtilities.exec("npm", ["publish", "--tag", tag.trim()], opts, callback);
  }

  static getExecOpts(directory, registry) {
    const opts = {
      cwd: directory,
      env: Object.assign({}, process.env)
    };

    if (registry) {
      opts.env = Object.assign({}, opts.env, {
        npm_config_registry: registry
      });
    }

    log.silly("getExecOpts", opts);
    return opts;
  }
}
