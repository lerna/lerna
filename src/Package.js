"use strict";

const dedent = require("dedent");
const log = require("npmlog");
const path = require("path");
const semver = require("semver");
const _ = require("lodash");

const dependencyIsSatisfied = require("./utils/dependencyIsSatisfied");
const NpmUtilities = require("./NpmUtilities");

class Package {
  constructor(pkg, location) {
    this._package = pkg;
    this._location = location;
  }

  get name() {
    return this._package.name;
  }

  get location() {
    return this._location;
  }

  get nodeModulesLocation() {
    return path.join(this._location, "node_modules");
  }

  get binLocation() {
    return path.join(this.nodeModulesLocation, ".bin");
  }

  get version() {
    return this._package.version;
  }

  set version(version) {
    this._package.version = version;
  }

  get bin() {
    return this._package.bin;
  }

  get dependencies() {
    return this._package.dependencies;
  }

  get devDependencies() {
    return this._package.devDependencies;
  }

  get peerDependencies() {
    return this._package.peerDependencies;
  }

  get allDependencies() {
    return Object.assign({}, this.devDependencies, this.dependencies);
  }

  get scripts() {
    return this._package.scripts || {};
  }

  set versionSerializer(versionSerializer) {
    this._versionSerializer = versionSerializer;

    if (versionSerializer) {
      this._package = versionSerializer.deserialize(this._package);
    }
  }

  isPrivate() {
    return !!this._package.private;
  }

  toJSON() {
    const pkg = _.cloneDeep(this._package);
    return this._versionSerializer ? this._versionSerializer.serialize(pkg) : pkg;
  }

  /**
   * Run a NPM script in this package's directory
   * @param {String} script NPM script to run
   * @param {Function} callback
   */
  runScript(script, callback) {
    log.silly("runScript", script, this.name);

    if (this.scripts[script]) {
      NpmUtilities.runScriptInDir(
        script,
        {
          args: [],
          directory: this.location,
          npmClient: "npm",
        },
        callback
      );
    } else {
      callback();
    }
  }

  /**
   * Run a NPM script synchronously in this package's directory
   * @param {String} script NPM script to run
   * @param {Function} callback
   */
  runScriptSync(script, callback) {
    log.silly("runScriptSync", script, this.name);

    if (this.scripts[script]) {
      NpmUtilities.runScriptInDirSync(
        script,
        {
          args: [],
          directory: this.location,
          npmClient: "npm",
        },
        callback
      );
    } else {
      callback();
    }
  }

  /**
   * Determine if a dependency version satisfies the requirements of this package
   * @param {Package} dependency
   * @param {Boolean} doWarn
   * @returns {Boolean}
   */
  hasMatchingDependency(dependency, doWarn) {
    log.silly("hasMatchingDependency", this.name, dependency.name);

    const expectedVersion = this.allDependencies[dependency.name];
    const actualVersion = dependency.version;

    if (!expectedVersion) {
      return false;
    }

    // check if semantic versions are compatible
    if (semver.satisfies(actualVersion, expectedVersion)) {
      return true;
    }

    if (doWarn) {
      log.warn(
        this.name,
        dedent`
          depends on "${dependency.name}@${expectedVersion}"
          instead of "${dependency.name}@${actualVersion}"
        `
      );
    }

    return false;
  }

  /**
   * Determine if a dependency has already been installed for this package
   * @param {String} depName Name of the dependency
   * @returns {Boolean}
   */
  hasDependencyInstalled(depName) {
    log.silly("hasDependencyInstalled", this.name, depName);

    return dependencyIsSatisfied(this.nodeModulesLocation, depName, this.allDependencies[depName]);
  }
}

module.exports = Package;
