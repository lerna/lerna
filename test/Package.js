import log from "npmlog";
// mocked modules
import NpmUtilities from "../src/NpmUtilities";
// helpers
import callsBack from "./helpers/callsBack";
import loggingOutput from "./helpers/loggingOutput";
// file under test
import Package from "../src/Package";

jest.mock("../src/NpmUtilities");

// silence logs
log.level = "silent";

describe("Package", () => {
  /* eslint no-underscore-dangle: ["error", { "allow": ["_package"] }] */
  let pkg;

  beforeEach(() => {
    pkg = new Package(
      {
        name: "my-package",
        version: "1.0.0",
        bin: "bin.js",
        scripts: { "my-script": "echo 'hello world'" },
        dependencies: { "my-dependency": "^1.0.0" },
        devDependencies: { "my-dev-dependency": "^1.0.0" },
        peerDependencies: { "my-peer-dependency": ">=1.0.0" },
      },
      "/path/to/package",
    );
  });

  describe("get .name", () => {
    it("should return the name", () => {
      expect(pkg.name).toBe("my-package");
    });
  });

  describe("get .location", () => {
    it("should return the location", () => {
      expect(pkg.location).toBe("/path/to/package");
    });
  });

  describe("get .version", () => {
    it("should return the version", () => {
      expect(pkg.version).toBe("1.0.0");
    });
  });

  describe("set .version", () => {
    it("should return the version", () => {
      pkg.version = "2.0.0";
      expect(pkg.version).toBe("2.0.0");
    });
  });

  describe("get .bin", () => {
    it("should return the bin", () => {
      expect(pkg.bin).toBe("bin.js");
    });
  });

  describe("get .dependencies", () => {
    it("should return the dependencies", () => {
      expect(pkg.dependencies).toEqual({ "my-dependency": "^1.0.0" });
    });
  });

  describe("get .devDependencies", () => {
    it("should return the devDependencies", () => {
      expect(pkg.devDependencies).toEqual({ "my-dev-dependency": "^1.0.0" });
    });
  });

  describe("get .peerDependencies", () => {
    it("should return the peerDependencies", () => {
      expect(pkg.peerDependencies).toEqual({ "my-peer-dependency": ">=1.0.0" });
    });
  });

  describe("get .allDependencies", () => {
    it("should return the combined dependencies", () => {
      expect(pkg.allDependencies).toEqual({
        "my-dependency": "^1.0.0",
        "my-dev-dependency": "^1.0.0",
      });
    });
  });

  describe("get .scripts", () => {
    it("should return the scripts", () => {
      expect(pkg.scripts).toEqual({
        "my-script": "echo 'hello world'",
      });
    });
  });

  describe(".isPrivate()", () => {
    it("should return if the package is private", () => {
      expect(pkg.isPrivate()).toBe(false);
    });
  });

  describe(".set versionSerializer", () => {
    it("should call 'deserialize' method of serializer'", () => {
      const mockSerializer = {
        serialize: jest.fn(obj => obj),
        deserialize: jest.fn(obj => obj),
      };

      pkg.versionSerializer = mockSerializer;

      expect(mockSerializer.deserialize).toBeCalled();
      expect(mockSerializer.deserialize).toBeCalledWith(pkg._package);
      expect(mockSerializer.serialize).not.toBeCalled();
    });
  });

  describe(".toJSON()", () => {
    it("should return clone of internal package for serialization", () => {
      expect(pkg.toJSON()).not.toBe(pkg._package);
      expect(pkg.toJSON()).toEqual(pkg._package);

      const implicit = JSON.stringify(pkg, null, 2);
      const explicit = JSON.stringify(pkg._package, null, 2);

      expect(implicit).toBe(explicit);
    });

    it("should not change internal package with versionSerializer", () => {
      pkg._package.state = "serialized";

      const mockSerializer = {
        serialize: jest.fn(obj => {
          obj.state = "serialized";
          return obj;
        }),
        deserialize: jest.fn(obj => {
          obj.state = "deserialized";
          return obj;
        }),
      };

      const serializedPkg = Object.assign({}, pkg._package, { state: "serialized" });
      const deserializedPkg = Object.assign({}, pkg._package, { state: "deserialized" });

      pkg.versionSerializer = mockSerializer;
      expect(mockSerializer.deserialize).toBeCalled();
      expect(pkg._package).toEqual(deserializedPkg);

      const serializedResult = pkg.toJSON();
      expect(pkg._package).toEqual(deserializedPkg);
      expect(serializedResult).toEqual(serializedPkg);

      expect(mockSerializer.serialize).toBeCalled();
    });

    it("should use versionSerializer.serialize on internal package before return", () => {
      const mockSerializer = {
        serialize: jest.fn(obj => obj),
        deserialize: jest.fn(obj => obj),
      };

      pkg.versionSerializer = mockSerializer;

      expect(pkg.toJSON()).toEqual(pkg._package);

      expect(mockSerializer.deserialize).toBeCalled();
      expect(mockSerializer.serialize).toBeCalled();
      expect(mockSerializer.serialize).toBeCalledWith(pkg._package);
    });
  });

  describe(".runScript()", () => {
    it("should run the script", done => {
      NpmUtilities.runScriptInDir = jest.fn(callsBack());

      pkg.runScript("my-script", () => {
        try {
          expect(NpmUtilities.runScriptInDir).lastCalledWith(
            "my-script",
            {
              args: [],
              directory: pkg.location,
              npmClient: "npm",
            },
            expect.any(Function),
          );

          done();
        } catch (ex) {
          done.fail(ex);
        }
      });
    });
  });

  describe(".runScriptSync()", () => {
    it("should run the script", () => {
      NpmUtilities.runScriptInDirSync = jest.fn(callsBack());

      pkg.runScriptSync("my-script", () => {});

      expect(NpmUtilities.runScriptInDirSync).lastCalledWith(
        "my-script",
        {
          args: [],
          directory: pkg.location,
          npmClient: "npm",
        },
        expect.any(Function),
      );
    });
  });

  describe(".hasMatchingDependency()", () => {
    it("should match included dependency", () => {
      expect(
        pkg.hasMatchingDependency({
          name: "my-dependency",
          version: "1.1.3",
        }),
      ).toBe(true);
    });

    it("should not match missing dependency", () => {
      expect(pkg.hasMatchingDependency({ name: "missing", version: "1.0.0" })).toBe(false);
      expect(loggingOutput()).toEqual([]);
    });

    it("should not match included dependency", () => {
      const result = pkg.hasMatchingDependency(
        {
          name: "my-dev-dependency",
          version: "2.0.7",
        },
        true,
      );

      expect(result).toBe(false);
      expect(loggingOutput()).toMatchSnapshot();
    });
  });
});
