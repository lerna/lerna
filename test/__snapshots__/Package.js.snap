// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Package .hasMatchingDependency() should not match included dependency 1`] = `
Array [
  "depends on \\"my-dev-dependency@^1.0.0\\"
instead of \\"my-dev-dependency@2.0.7\\"",
]
`;
